src/sdk/pynni/nni/analysis_utils/sensitivity/torch/sensitivity_analysis.py

```diff
@@ -0,0 +1,257 @@
+        analyze or parallelize the analysis process easily through
+        the %start and %end parameter.
+
+        Parameters
+        ----------
+            start:
+                Layer index of the sensitivity analysis start
+            end:
+                Layer index of the sensitivity analysis end
+            val_args:
+                args for the val_function
+            val_kwargs:
+                kwargs for the val_funtion
+                The val_funtion will be called as:
+                    val_function(*val_args, **val_kwargs)
+
+        Returns
+        -------
+            sensitivities:
+                dict object that stores the trajectory of the
+                accuracy when the prune ratio changes
+        """
+        if not end:
+            end = self.layers_count
+        assert start >= 0 and end <= self.layers_count
+        assert start <= end
+        if val_args is None:
+            val_args = []
+        if val_kwargs is None:
+            val_kwargs = {}
+        # Get the validation accuracy before pruning
+        if self.ori_acc is None:
+            self.ori_acc = self.val_func(*val_args, **val_kwargs)
+        namelist = list(self.target_layer.keys())
+        for layerid in range(start, end):
+            name = namelist[layerid]
+            self.sensitivities[name] = {}
+            for sparsity in self.sparsities:
+                # Calculate the actual prune ratio based on the already pruned ratio
+                sparsity = (
+                    1.0 - self.already_pruned[name]) * sparsity + self.already_pruned[name]
+                # TODO In current L1/L2 Filter Pruner, the 'op_types' is still necessary
+                # I think the L1/L2 Pruner should specify the op_types automaticlly
+                # according to the op_names
+                cfg = [{'sparsity': sparsity, 'op_names': [
+                    name], 'op_types': ['Conv2d']}]
+                pruner = self.Pruner(self.model, cfg)
+                pruner.compress()
+                val_acc = self.val_func(*val_args, **val_kwargs)
+                logger.info('Layer: %s Sparsity: %.2f Accuracy: %.4f',
+                            name, sparsity, val_acc)
+
+                self.sensitivities[name][sparsity] = val_acc
+                pruner._unwrap_model()
+                del pruner
+                # if the accuracy drop already reach the 'early_stop'
+                if val_acc + self.early_stop < self.ori_acc:
+                    break
+
+            # reset the weights pruned by the pruner, because
+            # out sparsities is sorted, so we donnot need to reset
+            # weight of the layer when the sparsity changes, instead,
+            # we only need reset the weight when the pruning layer changes.
+            self.model.load_state_dict(self.ori_state_dict)
+
+        return self.sensitivities
+
+    def visualization(self, outdir, merge=False):
```

Comment:
what is user interface for visualization?

Reply:
the 'visualization' is the function that visualizes the sensitivities of the model. This function should be called by user himself if he needs to visualize the sensitivity results. We already return the sensitivity results as a dict in function 'analysis', so the visualization is not set as a mandatory option.

For example:
s_analyzer = SensitivityAnalysis(model=net, val_func=val)
sensitivity = s_analyzer.analysis(val_args=[net])
s_analyzer.visualization(outdir, merge=True)